File: fluidresizelayout/src/main/java/me/saket/fluidresize/FluidContentResizer.java
package me.saket.fluidresize;

import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.app.Activity;
import android.view.ViewGroup;
import android.view.animation.Interpolator;

import androidx.interpolator.view.animation.FastOutSlowInInterpolator;

public final class FluidContentResizer {

    private ValueAnimator heightAnimator = new ObjectAnimator();
    private Interpolator interpolator = new FastOutSlowInInterpolator();

    public void listen(Activity activity) {
        ActivityViewHolder viewHolder = ActivityViewHolder.createFrom(activity);

        KeyboardVisibilityDetector.listen(viewHolder, visibilityChangedEvent -> animateHeight(viewHolder, visibilityChangedEvent));

        viewHolder.onDetach(() -> heightAnimator.cancel());
    }

    private void animateHeight(ActivityViewHolder viewHolder, KeyboardVisibilityChangedListener.KeyboardVisibilityChangedEvent event) {
        ViewGroup contentView = viewHolder.getContentView();

        setViewHeight(contentView, event.getContentHeightBeforeResize());

        heightAnimator.cancel();

        heightAnimator = ObjectAnimator.ofInt(event.getContentHeightBeforeResize(), event.getContentHeight());
        heightAnimator.setInterpolator(interpolator);
        heightAnimator.setDuration(300);
        heightAnimator.addUpdateListener(animation -> setViewHeight(contentView, (Integer) animation.getAnimatedValue()));
        heightAnimator.start();
    }

    private void setViewHeight(ViewGroup view, int height) {
        ViewGroup.LayoutParams lp = view.getLayoutParams();
        lp.height = height;
        view.setLayoutParams(lp);
    }
}

File: fluidresizelayout/src/main/java/me/saket/fluidresize/FluidContentResizer.kt
package me.saket.fluidresize

import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.app.Activity
import android.view.View
import android.view.ViewGroup
import android.view.Window
import androidx.interpolator.view.animation.FastOutSlowInInterpolator

object FluidContentResizer {

  private var heightAnimator: ValueAnimator = ObjectAnimator()

  fun listen(activity: Activity) {
    val viewHolder = findViews(activity)

    KeyboardVisibilityDetector.listen(viewHolder) {
      animateHeight(viewHolder, it)
    }

    viewHolder.onDetach {
      heightAnimator.cancel()
    }
  }

  private fun animateHeight(viewHolder: ActivityViewHolder, event: KeyboardVisibilityChanged) {
    val contentView = viewHolder.contentView
    contentView.setHeight(event.contentHeightBeforeResize)

    heightAnimator.cancel()

    heightAnimator = ObjectAnimator.ofInt(event.contentHeightBeforeResize, event.contentHeight).apply {
      interpolator = FastOutSlowInInterpolator()
      duration = 300
    }
    heightAnimator.addUpdateListener { contentView.setHeight(it.animatedValue as Int) }
    heightAnimator.start()
  }

  private fun View.setHeight(height: Int) {
    val params = layoutParams
    params.height = height
    layoutParams = params
  }

  /**
   * DecorView <- does not get resized, contains space for system Ui bars.
   * - LinearLayout <- does not get resized, contains space for action mode.
   * -- FrameLayout <- gets resized, contains space for action bar.
   * --- Activity content <- gets resized.
   */
  private fun findViews(activity: Activity): ActivityViewHolder {
    val decorView = activity.window.decorView as ViewGroup
    val decorViewChild = decorView.getChildAt(0) as ViewGroup
    val contentViewFrame = decorViewChild.getChildAt(1) as ViewGroup
    val contentView = decorView.findViewById<ViewGroup>(Window.ID_ANDROID_CONTENT)

    return ActivityViewHolder(
        decorView = decorView,
        contentViewFrame = contentViewFrame,
        contentView = contentView)
  }
}
